src/compiler/transformers/es2020.ts

```diff
@@ -159,12 +159,37 @@ namespace ts {
                 setOriginalNode(rightExpression, segment);
             }
 
-            const target = isDelete
-                ? createConditional(createNotNullCondition(leftExpression, capturedLeft, /*invert*/ true), createTrue(), createDelete(rightExpression))
-                : createConditional(createNotNullCondition(leftExpression, capturedLeft, /*invert*/ true), createVoidZero(), rightExpression);
+            let target;
+            const _isExpressionStatementDescendent = isExpressionStatementDescendent(node);
+            const notNull = createNotNullCondition(leftExpression, capturedLeft, /*invert*/ !_isExpressionStatementDescendent);
+            if (_isExpressionStatementDescendent) {
+                target = isDelete
+                    ? createLogicalAnd(notNull, createDelete(rightExpression))
+                    : createLogicalAnd(notNull, rightExpression);
+            }
+            else {
+                target = isDelete
+                    ? createConditional(notNull, createTrue(), createDelete(rightExpression))
+                    : createConditional(notNull, createVoidZero(), rightExpression);
+            }
+
             return thisArg ? createSyntheticReferenceExpression(target, thisArg) : target;
         }
 
+        function isExpressionStatementDescendent(node: Node): boolean {

```

Comment:
I'm not sure if this is the right approach but if it is I thought you may already have a utility function for it. If this is the right approach and there isn't an existing utility function I will move this to a utilities file.

Reply:
As this is a transform, it is unsafe to rely on `parent`. Tree transformations generate new syntax trees, but may reuse existing AST nodes. As these existing AST nodes must still participate in various types of resolution in the type checker, we cannot modify the `parent` pointer on these nodes. In addition, `parent` pointers are not set on synthetic nodes.

The best approach to this is to keep track of whether you are visiting a descendant of an `ExpressionStatement`, rather than relying on `parent`.

For example, in _es2018.ts_, we have a separate visitor callback we use when visiting a destructuring expression whose result is unused: 

https://github.com/microsoft/TypeScript/blob/611b77f2e64112786038a5925aeadd3bdd9a8b45/src/compiler/transformers/es2018.ts#L413-L420

https://github.com/microsoft/TypeScript/blob/611b77f2e64112786038a5925aeadd3bdd9a8b45/src/compiler/transformers/es2018.ts#L121-L128

We then pass that on when visiting a possible destructuring expression:

https://github.com/microsoft/TypeScript/blob/611b77f2e64112786038a5925aeadd3bdd9a8b45/src/compiler/transformers/es2018.ts#L166

This allows us to modify the emit behavior for an unused expression based on the tree we receive. As a result, if a future _es2021.ts_ transformation were to create a new tree containing a purely synthetic optional chain, the _es2020.ts_ transformer would still be able to perform this optimization even though the node is purely synthetic (and this has no `parent` pointer).

Reply:
Thank you! I will look into using a separate visitor callback.

Oddly enough I started writing an unrelated compiler a few days ago. I left it at this exact issue of a visitor needing unsafe reference to the tree structure, also with synthetic and intemediary nodes. Was unsure on the correct refactor until now. I will repay you! 😂